jwst/outlier_detection/tests/test_outlier_detection.py

```diff
@@ -251,74 +264,131 @@ def test_outlier_step_on_disk(we_three_sci, tmp_cwd):
         dm0.write(dm0.meta.filename)
 
     # Initialize inputs for the test based on filenames only
-    container = ModelContainer(filenames)
+    # needs to be an asn for ModelLibrary to load it in on_disk mode
+    asn = {
+    'asn_type': 'test',
+    'asn_id': 'o001',
+    'products': [
+        {
+            'name': 'product_a',
+            'members': [
+                {'expname': filenames[0], 'exptype': 'science'},
+                {'expname': filenames[1], 'exptype': 'science'},
+                {'expname': filenames[2], 'exptype': 'science'},
+            ]
+        },
+    ]}
+    return asn
+
+
+def test_outlier_step_on_disk(three_sci_as_asn, tmp_cwd):
+    """Test whole step with an outlier including saving intermediate and results files"""
+    container = ModelLibrary(three_sci_as_asn, on_disk=True)
+
+    # Save all the data into a separate array before passing into step
+    data_as_cube = container.map_function(lambda model, index: model.data, modify=False)
 
     result = OutlierDetectionStep.call(
-        container, save_results=True, save_intermediate_results=True
+        container, save_results=True, save_intermediate_results=True, in_memory=False
     )
 
     # Make sure nothing changed in SCI array
-    for image, corrected in zip(container, result):
-        np.testing.assert_allclose(image.data, corrected.data)
+    result.map_function(
+        lambda model, index: np.testing.assert_allclose(data_as_cube[index], model.data),
+        modify=False,
+    )
 
     # Verify source is not flagged
-    for r in result:
-        assert r.dq[7, 7] == datamodels.dqflags.pixel["GOOD"]
+    with result:
+        for r in result:
+            assert r.dq[7, 7] == datamodels.dqflags.pixel["GOOD"]
+            result.shelve(r, modify=False)
 
     # Verify CR is flagged
-    assert result[0].dq[12, 12] == OUTLIER_DO_NOT_USE
+    with result:
+        zeroth = result.borrow(0)
+        assert zeroth.dq[12, 12] == OUTLIER_DO_NOT_USE
+        result.shelve(zeroth, modify=False)
 
 
 def test_outlier_step_square_source_no_outliers(we_three_sci, tmp_cwd):
     """Test whole step with square source with sharp edges, no outliers"""
-    container = ModelContainer(list(we_three_sci))
+    container = ModelLibrary(list(we_three_sci))
 
     # put a square source in all three exposures
-    for ccont in container:
-        ccont.data[5:15, 5:15] += 1e3
-
-    pristine = container.copy()
-    result = OutlierDetectionStep.call(container)
+    with container:
+        for ccont in container:
+            ccont.data[5:15, 5:15] += 1e3
+            container.shelve(ccont)
+
+    # Save all the data into a separate array before passing into step
+    data_as_cube = []
+    dq_as_cube = []
+    with container:
+        for model in container:
+            data_as_cube.append(model.data)
+            dq_as_cube.append(model.dq)
+            container.shelve(model, modify=False)
```

Comment:
```suggestion
            data_as_cube.append(model.data.copy())
            dq_as_cube.append(model.dq.copy())
            container.shelve(model, modify=False)
```
Since the step is run with `in_memory=True` I believe the output results arrays will be the same arrays as the inputs. For the checks below this would mean (for example) `container[0].data is data_as_cube[0]` and the test could pass even if the step incorrectly flagged outliers. Adding the copies in this suggestion would protect against this (and mimics the previous `container.copy`).